File: examples/reactnative-basic/android/app/src/main/java/com/mapconductor/basic/ExampleGeoJSONStyler.kt
package com.mapconductor.basic

import android.content.Context
import android.graphics.Color
import android.net.Uri
import com.mapconductor.geojson.DefaultGeoJSONStyleProvider
import com.mapconductor.geojson.GeoJSONFeature
import com.mapconductor.geojson.GeoJSONStyleProviderInterface
import com.mapconductor.geojson.GeoJSONTileRenderer
import org.json.JSONArray
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.util.zip.ZipInputStream

/**
 * Example-only style provider: colors each railway line with the color declared in the
 * `*.style.json` packaged inside the same zip as the GeoJSON.
 *
 * This is the Android counterpart of `ios/MapConductorBasic/ExampleGeoJSONStyler.swift`
 * and of android-sdk's `example-app/.../geojson/layer/ExampleGeoJSONStyler.kt`. The point of
 * the sample is that **the application picks the colors**: `react-geojson-layer` only ships a
 * uniform fallback style, and the app registers a provider under an id which the JS side
 * references via `GeoJSONLayerState({ styleProviderId })`.
 */
class ExampleGeoJSONStyler(
    private val routeColors: Map<RouteKey, Int>,
) : GeoJSONStyleProviderInterface {
    override fun getStyle(
        feature: GeoJSONFeature,
        defaultStyle: GeoJSONTileRenderer.LayerStyle,
    ): GeoJSONTileRenderer.LayerStyle {
        val baseStyle = DefaultGeoJSONStyleProvider.getStyle(feature, defaultStyle)
        val companyName = feature.properties[COMPANY_PROPERTY]?.toString() ?: return baseStyle
        val lineName = feature.properties[LINE_PROPERTY]?.toString() ?: return baseStyle
        val color = routeColors[RouteKey(companyName, lineName)] ?: return baseStyle
        return baseStyle.copy(strokeColor = color)
    }

    data class RouteKey(
        val companyName: String,
        val lineName: String,
    )

    companion object {
        private const val COMPANY_PROPERTY = "N02_004"
        private const val LINE_PROPERTY = "N02_003"
        private const val STYLE_ENTRY = "N02-22_RailroadSection.style.json"

        /**
         * `sourceUri` is whatever the JS side handed to `<GeoJSONLayer sourceUri=... />`.
         * Expo's `Asset.localUri` gives a `file://` URL, so accept the same schemes the
         * bridge itself accepts in GeoJSONLayerRenderer.openUri().
         */
        fun fromZip(
            context: Context,
            sourceUri: String,
        ): ExampleGeoJSONStyler = ExampleGeoJSONStyler(parseRouteColors(readStyleJson(context, sourceUri)))

        private fun readStyleJson(
            context: Context,
            sourceUri: String,
        ): String {
            openUri(context, sourceUri).use { rawInput ->
                ZipInputStream(rawInput).use { zip ->
                    while (true) {
                        val entry = zip.nextEntry ?: break
                        val fileName = entry.name.substringAfterLast('/')
                        if (!entry.isDirectory && fileName.equals(STYLE_ENTRY, ignoreCase = true)) {
                            return zip.bufferedReader(Charsets.UTF_8).readText()
                        }
                        zip.closeEntry()
                    }
                }
            }
            throw IOException("$STYLE_ENTRY was not found in $sourceUri")
        }

        private fun openUri(
            context: Context,
            uriString: String,
        ): InputStream {
            val uri = Uri.parse(uriString)
            return when (uri.scheme?.lowercase()) {
                "content", "android.resource" ->
                    context.contentResolver.openInputStream(uri)
                        ?: throw IOException("Unable to open GeoJSON URI: $uriString")
                "file" -> FileInputStream(uri.path ?: throw IOException("file URI has no path: $uriString"))
                else -> FileInputStream(uriString)
            }
        }

        private fun parseRouteColors(json: String): Map<RouteKey, Int> {
            val result = mutableMapOf<RouteKey, Int>()
            val companies = JSONArray(json)
            for (companyIndex in 0 until companies.length()) {
                val company = companies.optJSONObject(companyIndex)?.optJSONObject("company") ?: continue
                val companyName = company.optString("name").takeIf(String::isNotBlank) ?: continue
                val lines = company.optJSONArray("lines") ?: continue
                for (lineIndex in 0 until lines.length()) {
                    val line = lines.optJSONObject(lineIndex) ?: continue
                    val lineName = line.optString("name").takeIf(String::isNotBlank) ?: continue
                    val color = line.optString("color").takeIf(String::isNotBlank) ?: continue
                    result[RouteKey(companyName, lineName)] = Color.parseColor(color)
                }
            }
            return result
        }
    }
}

File: examples/reactnative-basic/ios/MapConductorBasic/ExampleGeoJSONStyler.swift
import MapConductorGeoJSON
import MapConductorReactGeoJSONLayer
import UIKit

/// Example-only provider loaded from the style JSON packaged beside the GeoJSON.
final class ExampleGeoJSONStyler: GeoJSONStyleProvider {
  struct RouteKey: Hashable {
    let companyName: String
    let lineName: String
  }

  private let routeColors: [RouteKey: UIColor]

  init(routeColors: [RouteKey: UIColor]) {
    self.routeColors = routeColors
  }

  func style(
    for feature: GeoJSONFeature,
    defaultStyle: GeoJSONTileRenderer.LayerStyle
  ) -> GeoJSONTileRenderer.LayerStyle {
    let fallback = DefaultGeoJSONStyleProvider.shared.style(
      for: feature,
      defaultStyle: defaultStyle
    )
    guard let companyName = feature.properties[Self.companyProperty] as? String,
          let lineName = feature.properties[Self.lineProperty] as? String,
          let color = routeColors[RouteKey(companyName: companyName, lineName: lineName)] else {
      return fallback
    }
    return GeoJSONTileRenderer.LayerStyle(
      strokeColor: color,
      fillColor: fallback.fillColor,
      strokeWidth: fallback.strokeWidth,
      pointRadius: fallback.pointRadius
    )
  }

  static func fromZip(sourceUri: String) throws -> ExampleGeoJSONStyler {
    let data = try GeoJSONZipArchive.entryData(
      sourceUri: sourceUri,
      fileName: styleEntry
    )
    guard let companies = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
      throw StyleError.invalidJSON
    }
    var routeColors: [RouteKey: UIColor] = [:]
    for item in companies {
      guard let company = item["company"] as? [String: Any],
            let companyName = nonEmptyString(company["name"]),
            let lines = company["lines"] as? [[String: Any]] else { continue }
      for line in lines {
        guard let lineName = nonEmptyString(line["name"]),
              let colorString = nonEmptyString(line["color"]),
              let color = UIColor(androidHex: colorString) else { continue }
        routeColors[RouteKey(companyName: companyName, lineName: lineName)] = color
      }
    }
    return ExampleGeoJSONStyler(routeColors: routeColors)
  }

  private static func nonEmptyString(_ value: Any?) -> String? {
    guard let value = value as? String, !value.isEmpty else { return nil }
    return value
  }

  private enum StyleError: Error {
    case invalidJSON
  }

  private static let companyProperty = "N02_004"
  private static let lineProperty = "N02_003"
  private static let styleEntry = "N02-22_RailroadSection.style.json"
}

private extension UIColor {
  convenience init?(androidHex value: String) {
    let hex = value.hasPrefix("#") ? String(value.dropFirst()) : value
    guard hex.count == 6 || hex.count == 8,
          let parsed = UInt64(hex, radix: 16) else { return nil }
    let alpha: UInt64 = hex.count == 8 ? (parsed >> 24) & 0xff : 0xff
    let red = (parsed >> 16) & 0xff
    let green = (parsed >> 8) & 0xff
    let blue = parsed & 0xff
    self.init(
      red: CGFloat(red) / 255,
      green: CGFloat(green) / 255,
      blue: CGFloat(blue) / 255,
      alpha: CGFloat(alpha) / 255
    )
  }
}
